### PR TITLE
Pool oracle updates

### PR DESCRIPTION
**Pull request by aaoferreira**
Addressing audit findings

### CODE REVIEW

File: src/oracle/PoolOracle.sol [hunk 1]
```diff
@@ -46,15 +40,14 @@ contract PoolOracle is IPoolOracle {
...
-        uint256 windowSize_,
-        uint256 granularity_,
-        uint256 minTimeElapsed_
-    ) {
+    constructor(uint256 windowSize_, uint256 granularity_, uint256 minTimeElapsed_) {
         require(granularity_ > 1, "GRANULARITY");
-        require((periodSize = windowSize_ / granularity_) * granularity_ == windowSize_, "WINDOW_NOT_EVENLY_DIVISIBLE");
+
+        periodSize = windowSize_ / granularity_;
+        require(periodSize * granularity_ == windowSize_, "WINDOW_NOT_EVENLY_DIVISIBLE");
```
Comment (alcueca) [inline src/oracle/PoolOracle.sol:49]: Wouldn't this be a bit clearer?

`require(windowSize_ % granularity_ == 0, "WINDOW_NOT_EVENLY_DIVISIBLE");`

File: src/oracle/PoolOracle.sol [hunk 2]
```diff
@@ -240,26 +222,29 @@ contract PoolOracle is IPoolOracle {
...
+    {
         updateTime = block.timestamp;
         uint256 maturity = pool.maturity();
         if (updateTime >= maturity) {
             result = amount;
         } else {
             int128 price = _price(pool, peek(pool), g, maturity, updateTime);
-            result = amount.divu(WAD).div(price).mulu(WAD); // result = amount / price
+            require(price >= 0);
+            require(amount >> 192 == 0);
```
Comment (alcueca) [inline src/oracle/PoolOracle.sol:237]: Are you checking that `amount <= type(uint192).max`?

Comment (alcueca) [reply src/oracle/PoolOracle.sol:237]: Also, the `require` should have some message

File: src/oracle/PoolOracle.sol [hunk 3]
```diff
@@ -300,13 +286,11 @@ contract PoolOracle is IPoolOracle {
...
-        IPool pool,
-        uint256 twar,
-        int128 g,
-        uint256 maturity,
-        uint256 updateTime
-    ) internal view returns (int128 price) {
+    function _price(IPool pool, uint256 twar, int128 g, uint256 maturity, uint256 updateTime)
+        internal
+        view
+        returns (int128 price)
```
Comment (alcueca) [inline src/oracle/PoolOracle.sol:292]: I think that you need natspec stating clearly which parameters and return values are 64.64

File: src/oracle/PoolOracle.sol [hunk 4]
```diff
@@ -240,26 +222,29 @@ contract PoolOracle is IPoolOracle {
...
         updateTime = block.timestamp;
         uint256 maturity = pool.maturity();
         if (updateTime >= maturity) {
             result = amount;
         } else {
             int128 price = _price(pool, peek(pool), g, maturity, updateTime);
-            result = amount.divu(WAD).div(price).mulu(WAD); // result = amount / price
+            require(price >= 0);
+            require(amount >> 192 == 0);
+            result = (amount << 64) / uint128(price); // result = amount / price
```
Comment (alcueca) [inline src/oracle/PoolOracle.sol:238]: If the `<< 64` is intended to convert `amount` to 64.64, then it would be clearer to use `Math64x64.fromUint()`

**Review by alcueca — CHANGES_REQUESTED** [review]
Could you please add the audit in an `audits` folder at the root?

**Comment by aaoferreira** [comment]
>Could you please add the audit in an audits folder at the root?
We're waiting for the final report from ABDK, as soon as it's available I'll update the PR

File: src/oracle/PoolOracle.sol (hunk 2 above)
Comment (aaoferreira) [reply src/oracle/PoolOracle.sol:237]: This is a direct recommendation from ADBK's audit regarding gas savings. These requires already exists in their framework and are just being surfaced, so they're left as-is

File: src/oracle/PoolOracle.sol (hunk 4 above)
Comment (aaoferreira) [reply src/oracle/PoolOracle.sol:238]: Also a direct recommendation from ADBK's audit regarding gas savings. This provides a one operation to scale and divide and avoiding having to do another casting before dividing.

File: src/oracle/PoolOracle.sol (hunk 2 above)
Comment (alcueca) [reply src/oracle/PoolOracle.sol:237]: Cool, then please just add a comment saying so to redcue the head-scratching. Not everyone is going to be as smart as the ABDK guys.

File: src/oracle/PoolOracle.sol (hunk 4 above)
Comment (alcueca) [reply src/oracle/PoolOracle.sol:238]: Optimizooooors.

I like to reduce the surprise factor in code. So a little comment there, please.

**Review by alcueca — APPROVED** [review]
Please add those comments before you merge, and the audit in a later PR, please.

**Comment by alcueca** [comment]
Actually, put a comment at the top like:
`/// Audited by ABDK 13/03/23, report pending.`

**Comment by aaoferreira** [comment]
addressed the requests and will hold on the public announcement on our side to add the audit and merge this

**Pull request merged**